compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4125,13 +4135,14 @@ public void visit(BLangTypeInit cIExpr, AnalyzerData data) {
     }
 
     private BType checkObjectCompatibility(BType actualType, BLangTypeInit cIExpr, AnalyzerData data) {
-        switch (actualType.tag) {
+        BType referredType = Types.getReferredType(actualType);
```

Comment:
Since we are calling `checkObjectCompatibility()` recursively unwrapping the referred type if `actualType.tag` is a ReferenceTypeTag; why do we need to unwrap here?

Reply:
In the current master for `object type init`  `BType` is a `referenceType` but `determinedType` is `referred type`. This commit 39ebfdbfdce43615ec06c0d77e481754d5736846 will fix that issue. The `determinedType` is set when we do the type check in https://github.com/ballerina-platform/ballerina-lang/blob/618dc0ef34686f68e273015724d2b45246bd403c/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L4268

if we do not make this change `actualTypeInitType` will be `referred type`. Hence this fix

Reply:
I still don't get it. What if had the old `TypeTags.TYPEREFDESC` case in `checkObjectCompatibility()` and updated `getObjectConstructorReturnType` to a include a `TypeTags.TYPEREFDESC` case

Reply:
I have updated the mentioned change 9a3d2dd17eb2c3d79b31bb12aee715e106d99f0c